When restoring, resolve commandline passed sources against the current working directory

nkolev92 opened the pull request:
## Bug

Fixes: https://github.com/NuGet/Home/issues/9406
Regression: Yes  
* Last working version: 5.4 
* How are we preventing it in future: Better tests! Existing test was actually passing an absolute, not a relative path.

## Fix

Details: This got regressed when we did our static graph refactoring, where https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs was added. 

In particular, in non-static graph restore, we had to differentiate between the global properties passed in cmd and the properties set in the project files. The reason is that the props set in the project files need to resolved relative to the project. The cmd passed in props need to be resolved relative to the working directory. 
The concept of "Override" property was added in https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks/GetRestoreSettingsTask.cs. 

This just fixes that by explicitly referring to the global property in static graph and by passing the correct relative path in general. 

## Testing/Validation

Tests Added: Yes 
Reason for not adding tests:  
Validation:

nkolev92 committed:
Fix the relative source calculation

nkolev92 committed:
Use paths relative to the current working directory when dealing with global properties

nkolev92 committed:
add test

nkolev92 committed:
add tests

nkolev92 committed:
fix unit tests

nkolev92 committed:
fix static graph bug

nkolev92 committed:
fix test

nkolev92 committed:
fix test

nkolev92 commented:
🔔 @NuGet/nuget-client

kartheekp-ms commented on src/NuGet.Core/NuGet.Build.Tasks.Console/IMSBuildProject.cs:
```diff
@@ -21,6 +21,13 @@ internal interface IMSBuildProject : IMSBuildItem
+        /// </summary>
+        /// <param name="property"></param>
+        /// <returns>Trimmed global property value</returns>
+        string GetGlobalProperty(string property);
```
I think it would be better if the value is trimmed very top in the call stack. I prefer `GetGlobalPropertyValue` method name to be consistent with `MSBuildItemBase.GetPropertyValue`. Thoughts?

nkolev92 replied:
I think it is consistent with MSBuildItemBase.GetPropertyValue. 

https://github.com/NuGet/NuGet.Client/blob/8e4392fd2f766c2646a78917299c9ace6972db17/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildItemBase.cs#L19-L24

which has a doc from:
https://github.com/NuGet/NuGet.Client/blob/8e4392fd2f766c2646a78917299c9ace6972db17/src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/IMSBuildItem.cs#L18-L21

The first place that fetches it, trims it.

kartheekp-ms reviewed (commented):
I am still going through the changes and other docs to learn more about restore in general.

erdembayar commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildProjectInstance.cs:
```diff
@@ -50,5 +50,17 @@ public IEnumerable<IMSBuildItem> GetItems(string name)
+
+            if (string.IsNullOrWhiteSpace(value))
+            {
+                return null;
```
It looks GetPropertyValue return string.Empty if property is not found. 
So isn't better for keep consistency?

nkolev92 replied:
Where do you see that?
Can you please link it?

The implementation that I see is: https://github.com/NuGet/NuGet.Client/blob/c3219e49852add8371db25680a300cce5551ef00/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildItemBase.cs#L36-L39

When dealing with MSBuild empty is basically non existent.

erdembayar replied:
I just looked at [Line29](https://github.com/NuGet/NuGet.Client/blob/c3219e49852add8371db25680a300cce5551ef00/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildItemBase.cs#L29) then this one return empty string if property doesn't exist.

![image](https://user-images.githubusercontent.com/8766776/97762568-2ca83500-1ac6-11eb-8a9b-26ce1f16fecb.png)

nkolev92 replied:
Please **link** the code, not screenshots. 

That's also not our code. 

Our code returns `null` when the value is not set. 

The equivalent method for non-global properties is this one: 

https://github.com/NuGet/NuGet.Client/blob/c3219e49852add8371db25680a300cce5551ef00/src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildItemBase.cs#L21-L41

kartheekp-ms commented on src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs:
```diff
@@ -715,6 +717,55 @@ public static string[] GetSources(string projectDirectory, string[] sources, str
+                    values: additionalProjectSources,
+                    excludeValues: Enumerable.Empty<string>())
+                .Select(MSBuildRestoreUtility.FixSourcePath)
+                .ToArray();
```
I guess we can avoid calling ` .ToArray() ` if we can change the type of last parameter of `AppendItems` method to `IEnumerable<string>`

nkolev92 replied:
Fixed, but it doesn't net us a huge win since we now have to call `.Any()` which still allocates, but less than `ToArray()`

kartheekp-ms commented on src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs:
```diff
@@ -715,6 +717,55 @@ public static string[] GetSources(string projectDirectory, string[] sources, str
+            var filteredAdditionalProjectFallbackFolders = MSBuildRestoreUtility.AggregateSources(
+                    values: additionalProjectFallbackFolders,
+                    excludeValues: additionalProjectFallbackFoldersExcludes)
+                .ToArray();
```
I guess we can avoid calling ` .ToArray() ` if we can change the type of last parameter of `AppendItems` method to `IEnumerable<string>`

kartheekp-ms commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/BuildTasksUtilityTests.cs:
```diff
@@ -0,0 +1,71 @@
+            using (var testDir = TestDirectory.CreateInTemp())
+            {
+                // Arrange
+
```
nit: remove newline after arrange.

kartheekp-ms commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/BuildTasksUtilityTests.cs:
```diff
@@ -0,0 +1,71 @@
+            using (var testDir = TestDirectory.CreateInTemp())
+            {
+                // Arrange
+
```
nit: remove newline after arrange.

kartheekp-ms commented on src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs:
```diff
@@ -695,6 +696,7 @@ public static string[] GetFallbackFolders(string projectDirectory, string[] fall
             return AppendItems(projectDirectory, currentFallbackFolders, filteredAdditionalProjectFallbackFolders);
         }
 
+        [Obsolete("This method calculates the source overrides incorrectly and will be removed in a future release")]
```
The only difference I see between this `Obsolete` method and the new `GetSources` method implementation is addition of `startupDirectory` parameter and using its value to generate `sourcesOverride`. I would like to understand why new method was added instead of modifying the existing `GetSources()` method. The same comment is appliable to old `GetFallbackFolders` implementation that is marked as `Obsolete` in the PR.

nkolev92 replied:
Public API: https://github.com/NuGet/NuGet.Client/pull/3722/files#diff-aeed904ce6c0c0d7930e3ba06196b571bcdba15ef45c6299a9111a1a78457054

kartheekp-ms commented on test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs:
```diff
@@ -635,5 +635,63 @@ public async Task MsbuildRestore_WithStaticGraphAndRegularRestore_ErrorLoggedWhe
+
+                foreach (var asset in projectOutputPaths)
+                {
+                    new FileInfo(asset).Exists.Should().BeTrue(because: result.AllOutput);
```
As per [doc](https://docs.microsoft.com/en-us/dotnet/api/system.io.fileinfo.exists?view=netcore-3.1#remarks),  `FileInfo.Exists` When first called, FileInfo calls Refresh and caches information about the file. I think we can use `File.Exists` instead to avoid this cache operation.

nkolev92 replied:
We're not reusing the object, it doesn't matter.

kartheekp-ms commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/BuildTasksUtilityTests.cs:
```diff
@@ -0,0 +1,71 @@
+                     projectDirectory: projectDirectory,
+                     sources: new string[] { relativePath },
+                     sourcesOverride: null,
+                     additionalProjectSources: new string[] { },
```
[CA1825: Avoid zero-length array allocations](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1825). Use `Array.Empty<string>()` instead.

nkolev92 replied:
Fixed, but these are tests and when you write many tests, it's easier to have a variable string. 

After all, there's a reason why the analyzers are not raising these warnings in test projects :)

kartheekp-ms commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/BuildTasksUtilityTests.cs:
```diff
@@ -0,0 +1,71 @@
+                     projectDirectory: projectDirectory,
+                     sources: new string[] { relativePath },
+                     sourcesOverride: new string[] { relativePath },
+                     additionalProjectSources: new string[] { },
```
[CA1825: Avoid zero-length array allocations](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1825). Use `Array.Empty<string>()` instead.

kartheekp-ms reviewed (commented):
Additional thought - Do you think it will be nice to have a test where both `Source` and `FallbackSource` are overridden during restore?

nkolev92 reviewed (commented):
> Additional thought - Do you think it will be nice to have a test where both Source and FallbackSource are overridden during restore?

I think there's plenty of unit/integration tests handling them individually. 
Not sure what value would be there in overriding both at the same time.

nkolev92 committed:
feedback

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:
```diff
@@ -442,9 +442,10 @@ internal static string GetRestoreOutputPath(IMSBuildProject project)
+                project.GetGlobalProperty("OriginalMSBuildStartupDirectory"),
                 project.Directory,
                 project.SplitPropertyValueOrNull("RestoreSources"),
-                project.SplitPropertyValueOrNull("RestoreSourcesOverride"),
+                project.SplitGlobalPropertyValueOrNull("RestoreSources"),
```
What's the difference between a global property and a (not global) property?

Is it worthwhile calling the method once before the return statement, and pass the same result to the two different parameters, or are the return values of the two methods different?

nkolev92 replied:
Global properties are ones that are passed on the commandline, and when they are relative, they should be resolved wrt to the working directory. 
non global properties are the effective ones, and unless they are global properties explicitly, when relative, they should be calculated with respect to the project file. 

The props could be different. In fact the most common case is that there's no global property.

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:
```diff
@@ -765,9 +767,10 @@ private PackageSpec GetPackageSpec(IMSBuildProject project, IReadOnlyDictionary<
+                        project.GetProperty("MSBuildStartupDirectory"),
                         project.Directory,
                         project.SplitPropertyValueOrNull("RestoreFallbackFolders"),
-                        project.SplitPropertyValueOrNull("RestoreFallbackFoldersOverride"),
+                        project.SplitGlobalPropertyValueOrNull("RestoreFallbackFolders"),
```
like my other comment, I don't understand the difference between the two calls for the same property name.

nkolev92 replied:
Hope https://github.com/NuGet/NuGet.Client/pull/3722#discussion_r517617717 covers it.

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs:
```diff
@@ -709,15 +710,61 @@ public static string[] GetSources(string projectDirectory, string[] sources, str
         {
-            if (additional == null || additional.Length == 0)
+            // Sources
+            var currentSources = RestoreSettingsUtils.GetValue(
```
offtopic: `RestoreSettingsUtils.GetValue<T>(Func<T>)` is a poorly named, and confusing method.

nkolev92 replied:
If we remove the project from the public API(NuGet.Build.Tasks), you can change it :D

zivkan commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.RestoreEx.targets:
```diff
@@ -31,6 +31,7 @@ Copyright (c) .NET Foundation. All rights reserved.
         RestorePackagesConfig="$(RestorePackagesConfig)"
         SolutionPath="$(SolutionPath)"
         ProcessFileName="$(NuGetConsoleProcessFileName)"
+        MSBuildStartupDirectory="$(MSBuildStartupDirectory)"
```
Does this mean that when our task is executing, `Directory.CurrentDirectory()` does not return the directory where the user ran msbuild?

Converting relative to absolute paths as soon as possible and relying on absolute paths throughout our codebase seems like a good idea. But it just seems weird that we can't use the runtime's built-in methods.

nkolev92 replied:
It does not, because the working directory is going to the NuGet assembly that started this new process. 

Relying on absolute codepaths is nice, but this allows us to keep the code for settings resolution as *close* as possible.
Furthermore the current NuGet process is already a `forwarder` without much of it's own restore logic.

zivkan commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/MSBuildStaticGraphRestoreTests.cs:
```diff
@@ -221,8 +221,11 @@ public void GetPackagesPath_WhenPathOrOverrideSpecified_PathIsCorrect(string pac
-                    ["RestorePackagesPathOverride"] = packagesPathOverride,
                     ["RestorePackagesPath"] = packagesPath
+                },
+                new Dictionary<string, string>
```
use named parameters. I have no idea what the difference between the two dictionaries are.

zivkan commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/MSBuildStaticGraphRestoreTests.cs:
```diff
@@ -505,8 +508,11 @@ public void GetRepositoryPath_WhenPathSolutionOrOverrideSpecified_CorrectPathDet
-                    ["RestoreRepositoryPathOverride"] = repositoryPathOverride,
                     ["SolutionPath"] = solutionPath == null || solutionPath == "*Undefined*" ? solutionPath : UriUtility.GetAbsolutePath(testDirectory, solutionPath)
+                },
+                new Dictionary<string, string>
```
named parameters, please.

zivkan commented on test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/MSBuildStaticGraphRestoreTests.cs:
```diff
@@ -607,7 +613,10 @@ public void GetSources_WhenRestoreSourcesAndRestoreSourcesOverrideSpecified_Corr
                 ["RestoreSources"] = "https://source1;https://source2",
-                ["RestoreSourcesOverride"] = "https://source3"
+            },
+            new Dictionary<string, string>
```
named parameters, please.

nkolev92 committed:
address feedback

nkolev92 commented:
@zivkan 
I addressed the code feedback, and answered your questions. 

Given that you approved, I'm planning on merging when the CI passes. 
I'd be happy to continue the discussion for the questions you have and address things in future PRs if necessary. 

Let me know if you disagree.

dfederm commented on src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:
```diff
@@ -606,6 +607,8 @@ private DependencyGraphSpec GetDependencyGraphSpec(string entryProjectPath, IDic
                     // Get the PackageSpecs in parallel because creating each one is relatively expensive so parallelism speeds things up
                     Parallel.ForEach(projects, new ParallelOptions { MaxDegreeOfParallelism = Environment.ProcessorCount }, project =>
                     {
+                        MSBuildLogger.LogMinimal($"For {project.OuterProject.FullPath}," +
```
@nkolev92 This seems to spam the console, even with minimal logging is specified. Did you mean to leave this in there?

nkolev92 replied:
Not on purpose. 
I'll fix it.
